patches/server/0878-More-Teleport-API.patch

```diff
@@ -0,0 +1,168 @@
+@@ -560,15 +560,33 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
+ 
+     @Override
+     public boolean teleport(Location location, TeleportCause cause) {
++        // Paper start - Teleport passenger API
++        return teleport(location, cause, false);
++    }
++
++    @Override
++    public boolean teleport(Location location, TeleportCause cause, boolean ignorePassengers, boolean dismount) {
++        // Paper end
+         Preconditions.checkArgument(location != null, "location");
+         location.checkFinite();
++        // Paper start - Teleport passenger API
++        // Don't allow teleporting between worlds while keeping passengers
++        if (ignorePassengers && entity.isVehicle() && location.getWorld() != this.getWorld()) {
++            return false;
++        }
++
++        // Don't allow to teleport between worlds if remaining on vehicle
++        if (!dismount && entity.isPassenger() && location.getWorld() != this.getWorld()) {
++            return false;
++        }
++        // Paper end
+ 
+-        if (this.entity.isVehicle() || this.entity.isRemoved()) {
++        if ((!ignorePassengers && this.entity.isVehicle()) || this.entity.isRemoved()) { // Paper - Teleport passenger API
+             return false;
+         }
+ 
+         // If this entity is riding another entity, we must dismount before teleporting.
+-        this.entity.stopRiding();
++        if (dismount) this.entity.stopRiding(); // Paper - Teleport passenger API
+ 
+         // Let the server handle cross world teleports
+         if (!location.getWorld().equals(this.getWorld())) {
+diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+index 6feeadad9aecb7d63e24d5daae115a93e39aeb3d..d26bb0032ca0729e68a3418d1c6be84355e63488 100644
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+@@ -1040,8 +1040,66 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
+ 
+     @Override
+     public boolean teleport(Location location, PlayerTeleportEvent.TeleportCause cause) {
++        // Paper start - Teleport API
++        return teleport(location, cause, false);
++    }
++
++    @Override
++    public boolean teleport(Location location, PlayerTeleportEvent.TeleportCause cause, boolean ignorePassengers, boolean dismount) {
++        return teleport(location, cause, ignorePassengers, dismount, new io.papermc.paper.entity.RelativeTeleportFlag[0]);
++    }
++
++    @Override
++    public void lookAt(@NotNull org.bukkit.entity.Entity entity, @NotNull io.papermc.paper.entity.LookAnchor playerAnchor, @NotNull io.papermc.paper.entity.LookAnchor entityAnchor) {
++        net.minecraft.commands.arguments.EntityAnchorArgument.Anchor[] anchors = net.minecraft.commands.arguments.EntityAnchorArgument.Anchor.values();
```

Comment:
You could cache this array statically to avoid copying the `Anchor` values each time this method is called.

Reply:
Could also just add a simple switch statement converted, similar to `toApiRelativeFlag` or `toNmsRelativeFlag`, which would a) avoid the 2-pointer array allocation caused by calling #values() and b) removes the dependency on the enums to be ordered alike.

Reply:
Done